cmd/scollector/collectors/rabbitmq.go

```diff
@@ -0,0 +1,440 @@
+package collectors
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"bytes"
+
+	"bosun.org/metadata"
+	"bosun.org/opentsdb"
+)
+
+var (
+	rmqQueueStatusMap = map[string]int{
+		"running": 0,
+		"syncing": 1,
+		"flow":    2,
+		"idle":    3,
+		"down":    4,
+	}
+)
+
+const (
+	defaultRabbitmqURL string = "http://guest:guest@localhost:15672"
+	rmqPrefix                 = "rabbitmq."
+)
+
+func init() {
+	collectors = append(collectors, &IntervalCollector{F: c_rabbitmq_overview, Enable: enableRabbitmq})
+	collectors = append(collectors, &IntervalCollector{F: c_rabbitmq_queues, Enable: enableRabbitmq})
+	collectors = append(collectors, &IntervalCollector{F: c_rabbitmq_nodes, Enable: enableRabbitmq})
+}
+
+// RabbitMQ registers a RabbitMQ collector.
+func RabbitMQ(url string) error {
+	safeURL, err := urlUserHost(url)
+	if err != nil {
+		return err
+	}
+	collectors = append(collectors,
+		&IntervalCollector{
+			F: func() (opentsdb.MultiDataPoint, error) {
+				return rabbitmqOverview(url)
+			},
+			name: fmt.Sprintf("rabbitmq-overview-%s", safeURL),
+		},
+		&IntervalCollector{
+			F: func() (opentsdb.MultiDataPoint, error) {
+				return rabbitmqNodes(url)
+			},
+			name: fmt.Sprintf("rabbitmq-nodes-%s", safeURL),
+		},
+		&IntervalCollector{
+			F: func() (opentsdb.MultiDataPoint, error) {
+				return rabbitmqQueues(url)
+			},
+			name: fmt.Sprintf("rabbitmq-queues-%s", safeURL),
+		})
+	return nil
+}
+
+func enableRabbitmq() bool {
+	return enableURL(defaultRabbitmqURL)()
+}
+
+func c_rabbitmq_overview() (opentsdb.MultiDataPoint, error) {
+	return rabbitmqOverview(defaultRabbitmqURL)
+
+}
+func c_rabbitmq_nodes() (opentsdb.MultiDataPoint, error) {
+	return rabbitmqNodes(defaultRabbitmqURL)
+}
+func c_rabbitmq_queues() (opentsdb.MultiDataPoint, error) {
+	return rabbitmqQueues(defaultRabbitmqURL)
+}
+
+func rabbitmqOverview(s string) (opentsdb.MultiDataPoint, error) {
+	p := rmqPrefix + "overview."
+	res, err := http.Get(s + "/api/overview")
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+	content, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+	var o rmqOverview
+	if err := json.NewDecoder(bytes.NewReader(content[:])).Decode(&o); err != nil {
+		return nil, err
+	}
+	var md opentsdb.MultiDataPoint
+	host := strings.Split(o.Node, "@")[1]
```

Comment:
Please check the result of split here. If something goes wrong and there is no @, this will panic.